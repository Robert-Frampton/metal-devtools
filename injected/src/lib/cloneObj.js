export const ITERABLE_KEY = '@@__IMMUTABLE_ITERABLE__@@';

// Copied pattern from MDN, https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
function cloneObj(objectToBeCloned, visited = new Set()) {
	if (!(objectToBeCloned instanceof Object)) {
		return objectToBeCloned;
	}

	let objectClone;

	try {
		if (objectToBeCloned && objectToBeCloned[ITERABLE_KEY]) {
			objectToBeCloned = objectToBeCloned.toJS();
		}

		const Constructor = objectToBeCloned.constructor;

		switch (Constructor) {
			case RegExp:
				objectClone = new Constructor(objectToBeCloned);
				break;
			case Date:
				objectClone = new Constructor(objectToBeCloned.getTime());
				break;
			case HTMLElement:
				objectClone = `<${objectToBeCloned.tagName} />`;
				break;
			case Function:
				if (objectToBeCloned.name) {
					objectClone = `function ${objectToBeCloned.name}()`;
				}
				else if (objectToBeCloned.__jsxDOMWrapper) {
					objectClone = '<JSXElement />';
				}
				else {
					objectClone = 'function()';
				}
				break;
			default:
				objectClone = new Constructor();
		}
	} catch (err) {
		console.log('%c metal-devtools extension: (`clone`)\n', 'background: rgb(136, 18, 128); color: #DDD', err);
		console.log('%c Args:', 'background: rgb(136, 18, 128); color: #DDD', objectToBeCloned);
	}

	if (objectClone instanceof Object) {
		visited.add(objectToBeCloned);

		for (let key in objectToBeCloned) {
			const prop = objectToBeCloned[key];

			if (typeof prop !== 'undefined') {
				objectClone[key] = visited.has(prop) ? '[Circular]' : cloneObj(prop, visited);
			}
		}
	}

	return objectClone;
}

export default cloneObj;
