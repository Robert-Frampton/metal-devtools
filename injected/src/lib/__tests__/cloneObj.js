jest.unmock('../cloneObj');

import Component from 'metal-jsx'; // eslint-disable-line
import {List, Map} from 'immutable';

import cloneObj, {ITERABLE_KEY} from '../cloneObj';

describe('cloneObj', () => {
	test('should return item if not an object', () => {
		const bool = true;
		const num = 123;
		const str = 'foo';

		expect(cloneObj(bool)).toEqual(bool);
		expect(cloneObj(num)).toEqual(num);
		expect(cloneObj(str)).toEqual(str);
	});

	test('should copy anonymous function', () => {
		expect(cloneObj(() => {})).toEqual('function()');
	});

	test('should copy named function', () => {
		function fooFn() {};

		expect(cloneObj(fooFn)).toEqual('function fooFn()');
	});

	test('should copy array', () => {
		expect(cloneObj([])).toBeInstanceOf(Array);
	});

	test('should copy Date', () => {
		expect(cloneObj(new Date())).toBeInstanceOf(Date);
	});

	test('should copy DOM node', () => {
		const node = {
			constructor: HTMLElement,
			tagName: 'DIV'
		};

		expect(cloneObj(node)).toEqual('<DIV />');
	});

	test('should copy immutable type to be plain js', () => {
		expect(cloneObj(List())).toBeInstanceOf(Array);
		expect(cloneObj(Map())).toBeInstanceOf(Object);
	});

	test('should copy jsx type to be function string', () => {
		expect(cloneObj(<div></div>)).toEqual('<JSXElement />');
	});

	test('should copy object', () => {
		expect(cloneObj({foo: 'bar'})).toBeInstanceOf(Object);
	});

	test('should copy nested object', () => {
		const nestedObj = {
			foo: {
				bar: 'baz'
			}
		};

		expect(cloneObj(nestedObj)).toBeInstanceOf(Object);
	});

	test('should copy RegExp', () => {
		expect(cloneObj(/test/)).toBeInstanceOf(RegExp);
	});

	test('should throw console.log on error', () => {
		const initialLog = console.log;

		console.log = jest.fn();

		cloneObj({
			[ITERABLE_KEY]: true,
			toJS: () => {
				throw new Error();
			}
		});

		expect(console.log).toHaveBeenCalledTimes(2);

		console.log = initialLog;
	});

	test.only('should not follow circular references', () => {
		const objA = {};
		const objB = {};
		const objC = {};
		objA.ref = objB;
		objB.ref = objC;
		objC.ref = objA;

		expect(cloneObj(objA)).toMatchObject(
			{
				ref: {
					ref: {
						ref: '[Circular]'
					}
				}
			}
		);
	});
});
