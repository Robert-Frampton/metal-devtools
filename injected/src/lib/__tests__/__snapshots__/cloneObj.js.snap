exports[`cloneObj should copy anonymous function 1`] = `
Object {
  "__metal_devtools_read_only": true,
  "value": "function()",
}
`;

exports[`cloneObj should copy jsx type to be function string 1`] = `
Object {
  "__jsxDOMWrapper": true,
  "__metal_devtools_read_only": true,
  "value": "<JSXElement />",
}
`;

exports[`cloneObj should copy named function 1`] = `
Object {
  "__metal_devtools_read_only": true,
  "value": "fooFn()",
}
`;

exports[`cloneObj should follow redundant-acyclic references 1`] = `
Object {
  "left": Object {
    "ref": Object {
      "name": "leaf",
    },
  },
  "right": Object {
    "ref": Object {
      "name": "leaf",
    },
  },
}
`;

exports[`cloneObj should not follow circular references 1`] = `
Object {
  "ref": Object {
    "ref": Object {
      "ref": "[Circular]",
    },
  },
}
`;

exports[`cloneObj should return item if not an object 1`] = `true`;

exports[`cloneObj should return item if not an object 2`] = `123`;

exports[`cloneObj should return item if not an object 3`] = `"foo"`;

exports[`cloneObj should use constructor name 1`] = `"mockConstructor"`;
